Rework CB management during stacking context tree construction

mrobinson opened the pull request:
Manage containing blocks and WebRender `SpaceAndClip` during stacking context tree construction using the `ContainingBlockInfo` data structure. This will allow us to reuse this data structure whenever we traverse the fragment tree. In addition, `StackingContextBuilder` is no longer necessary at all. This change also fixes some bugs where fixed position fragments were not placed in the correct spatial node. Unfortunately, these fixes are difficult to test because of #29659.

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes.

mrobinson committed:
Rework CB management during stacking context tree construction

Manage containing blocks and WebRender SpaceAndClip during stacking
context tree constuction using the ContainingBlockInfo data structure.
This will allow us to reuse this data structure whenever we traverse the
fragment tree. In addition, StackingContextBuilder is no longer
necessary at all. This change also fixes some bugs where fixed position
fragments were not placed in the correct spatial node. Unfortunately,
these fixes are difficult to test because of #29659.

mrobinson commented:
I should be able to write a test for this once #29661 lands.

mrobinson commented:
@bors-servo try=wpt-2020

bors-servo commented:
:hourglass: Trying commit 90f3cd67d0ebaad2e08438171fa243b02ffa153f with merge 581314db7df83071df25fcd1df457560c3b64603...

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#39712) with upstreamable changes.

bors-servo commented:
:broken_heart: Test failed - [checks-github](https://github.com/servo/servo/actions/runs/4810615874/jobs/8564574269)

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#39712).

mrobinson commented:
@bors-servo try=wpt-2020

bors-servo commented:
:hourglass: Trying commit 44021d372b22d94b701ed2d1ed2edc57b3a62ec2 with merge 1eccb47acd290f3723b4b1aa9a7168b896c74423...

bors-servo commented:
:broken_heart: Test failed - [checks-github](https://github.com/servo/servo/actions/runs/4811361343/jobs/8566026491)

mrobinson commented:
These failed tests were fixed by #29671.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#39712).

mrobinson commented:
I've updated the PR with the following changes:
- There is now an assertion in `BoxFragment::build_stacking_context_tree_maybe_creating_reference_frame` verifying that all fragments that establish a reference frame also establish a containing block for all descendants. This is to help show that the temporarily bogus properties on `ContainingBlockInfo` will be overwritten before they are read.
- There is a comment when starting the stacking context tree construction explaining why `ContainingBlockInfo` is initialized with 3 containing blocks instead of using one of the `new_` functions.
- I have renamed a variable to make it consistent with another in the same function.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#39712).

delan commented:
Great work, and thanks again for walking me through it!

@bors-servo r+

bors-servo commented:
:pushpin: Commit 33c7963 has been approved by `delan`

<!-- @bors-servo r=delan 33c7963f00a80d6b3b081fc0f180a67254bee4bb -->

bors-servo commented:
:hourglass: Testing commit 33c7963f00a80d6b3b081fc0f180a67254bee4bb with merge 9d0245ab47d242403fcdf130ad2f74145cd38dae...

mrobinson commented:
> -  FAIL [expected PASS] /css/css-transforms/transform-containing-block-and-scrolling-area-for-fixed.html

Odd, this test works locally. I wonder if this is an issue with flakiness in scrolling.

mrobinson commented:
@bors-servo try=wpt-2020

bors-servo commented:
:hourglass: Trying commit 95e7099637a3a5084cda807090b44a2e54aa708b with merge c19b40cfae44eeead364c58a9c19f291d22910b3...

mrobinson commented:
> Odd, this test works locally. I wonder if this is an issue with flakiness in scrolling.

Oh, obviously the new test will not pass in the old layout system so needs proper expectations!

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#39712).

mrobinson commented:
@bors-servo r=delan

bors-servo commented:
:hourglass: Testing commit 9038db80655392f277e1bdaedb26f76ad2f320b8 with merge c0938982182754ac051698bed9da4a3f08eb49e5...

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#39712).

mrobinson commented:
@bors-servo r=delan

- Rebased.

bors-servo commented:
:pushpin: Commit 0c13fcb has been approved by `delan`

<!-- @bors-servo r=delan 0c13fcb9f27a51c787a48cca6df91850d7a9fc74 -->

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#39712).

bors-servo commented:
:hourglass: Testing commit 0c13fcb9f27a51c787a48cca6df91850d7a9fc74 with merge 77a184a0e7379a63a0d9d8bf442d8dd5c3b5e307...

bors-servo commented:
:sunny: Test successful - [checks-github](https://github.com/servo/servo/actions/runs/4870812940/jobs/8689096063)
Approved by: delan
Pushing 77a184a0e7379a63a0d9d8bf442d8dd5c3b5e307 to master...
